### PR TITLE
Added new endpoint: Grafana Dashboard objects

### PR DESCRIPTION
**Pull request by etiennejournet**
Hello,

Here is a PR that adds the client for GET/DELETE/CREATE|UPDATE on grafana dashboards. [Linked to this.](https://github.com/logzio/terraform-provider-logzio/issues/95#issuecomment-974768220)

Looking forward to a review from you since I'm still learning Go :)

A question about the dashboards `uid` field : I treat dashboards as map[string]interface{} because I feel like it would be too complicated to declare all fields. But having the `uid` field in the client response object might be helpful, specially with the terraform provider. What do you think ?

Thanks

### CODE REVIEW

**Comment by Doron-Bargo** [comment]
Thanks @etiennejournet for your contribution
We will review it shortly

File: grafana_objects/client_grafana_objects.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+import (
+	"fmt"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/logzio/logzio_terraform_client/client"
+)
+
+const (
+	grafanaObjectsDashboardsByUID        = "%s/v1/grafana/api/dashboards/uid/%s"
+	grafanaObjectsDashboardsCreateUpdate = "%s/v1/grafana/api/dashboards/db"
```
Comment (mirii1994) [inline grafana_objects/client_grafana_objects.go:12]:
```suggestion
	grafanaObjectServiceEndpoint        = "%s/v1/grafana/api/dashboards"
```

Please add a `grafanaObjectServiceEndpoint` const that will have the base service endpoint.
Then, change the existing 2 consts to use it and move them to the file where it's used (grafanaObjectsDashboardsByUID to the get file, grafanaObjectsDashboardsCreateUpdate to the create update file).
You can see an example here:
https://github.com/logzio/logzio_terraform_client/blob/master/archive_logs/client_archive_logs.go#L8
and here:
https://github.com/logzio/logzio_terraform_client/blob/master/archive_logs/client_archive_logs_retrieve.go#L11

File: grafana_objects/client_grafana_objects_create_update.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+	"github.com/logzio/logzio_terraform_client/client"
+)
+
+type CreateUpdatePayload struct {
+	Dashboard map[string]interface{} `json:"dashboard"`
+	FolderId  int                    `json:"folderId"`
+	FolderUid int                    `json:"folderUid"`
+	Message   string                 `json:"message"`
+	Overwrite bool                   `json:"overwrite"`
+}
```
Comment (mirii1994) [inline grafana_objects/client_grafana_objects_create_update.go:20]: First, please move the structs that represent a request and a response to `client_grafana_objects.go`.
Second, ss a general rule of thumb, if you can use structs to represent your JSON data, you should use them. The only good reason to use maps would be if it were not possible to use structs due to the uncertain nature of the keys or values in the data.
Please change the map, and create a dashboard object with the relevant fields, then use it as a nested field in CreateUpdatePayload etc...

File: grafana_objects/client_grafana_objects_create_update.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+	jsonBytes, err = ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("could not read response body: %w", err)
+	}
+
+	if !logzio_client.CheckValidStatus(resp, []int{http.StatusOK}) {
+		return nil, fmt.Errorf("%d %s", resp.StatusCode, jsonBytes)
+	}
+
```
Comment (mirii1994) [inline grafana_objects/client_grafana_objects_create_update.go:62]: Whenever you initiate a call to the Logzio api, please use the `CallLogzioApi` function instead.
You can see an example here:
https://github.com/logzio/logzio_terraform_client/blob/master/archive_logs/client_archive_logs_retrieve.go#L19

File: grafana_objects/client_grafana_objects_create_update_test.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"testing"
+
+	"github.com/logzio/logzio_terraform_client/grafana_objects"
+	"github.com/stretchr/testify/assert"
+)
+
+func createUpdateMockHandler(t *testing.T) func(http.ResponseWriter, *http.Request) {
```
Comment (mirii1994) [inline grafana_objects/client_grafana_objects_create_update_test.go:14]: Please refactor the test like in here:
https://github.com/logzio/logzio_terraform_client/blob/master/archive_logs/archive_logs_setup_test.go

File: grafana_objects/testdata/createupdate.json [hunk 5]
```diff
@@ -0,0 +1,10 @@
+{
```
Comment (mirii1994) [inline grafana_objects/testdata/createupdate.json:1]: Please add directory "fixture" under the directory "testdata", and place the files there.

**Review by mirii1994 — CHANGES_REQUESTED** [review]
Hi @etiennejournet , thank you so much for your contribution, great job!
Please see my comments :)

File: grafana_objects/client_grafana_objects_create_update_test.go (hunk 4 above)
Comment (etiennejournet) [reply grafana_objects/client_grafana_objects_create_update_test.go:14]: hi @mirii1994 , thanks for your review, do you need me to do it for get_by_uid and delete_by_uid too?

File: grafana_objects/client_grafana_objects.go (hunk 1 above)
Comment (etiennejournet) [reply grafana_objects/client_grafana_objects.go]: done

File: grafana_objects/client_grafana_objects_create_update.go (hunk 2 above)
Comment (etiennejournet) [reply grafana_objects/client_grafana_objects_create_update.go]: done, not sure abut `Test_ValidateTimepickerStruct`, `Test_ValidateTemplatingStruct` and `Test_ValidateDashboardObjectsStruct` though

File: grafana_objects/testdata/createupdate.json (hunk 5 above)
Comment (etiennejournet) [reply grafana_objects/testdata/createupdate.json]: done

File: grafana_objects/client_grafana_dashboard_structs.go [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+	FolderUid string `json:"folderUid"`
+	Slug      string `json:"slug"`
+}
+
+type DashboardObject struct {
+	Id            int            `json:"id"`
+	Uid           string         `json:"uid"`
+	Title         string         `json:"title`
+	Tags          []string       `json:"tags"`
+	Style         string         `json:"style"`
```
Comment (mirii1994) [inline grafana_objects/client_grafana_dashboard_structs.go:16]: Some of those fields do not appear in the [documentation](https://docs.logz.io/api/#operation/createDashboard) (Style, Editable, GraphToolTip, Time, Timepicker, Templating, Annotations, Panels). Why did you add those?

File: grafana_objects/client_grafana_objects_create_update.go (hunk 3 above)
Comment (mirii1994) [reply grafana_objects/client_grafana_objects_create_update.go:49]: Please apply this ⬆️

File: grafana_objects/client_grafana_objects_delete_by_uid.go [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+	"net/http"
+
+	logzio_client "github.com/logzio/logzio_terraform_client"
+	"github.com/logzio/logzio_terraform_client/client"
+)
+
+// Delete allows deleting Grafana objects configuration.
+// https://docs.logz.io/api/#operation/deleteDashboarById
+func (c *GrafanaObjectsClient) Delete(objectUid string) (*DeleteResults, error) {
+	url := fmt.Sprintf(grafanaObjectsDashboardsByUID, c.BaseUrl, objectUid)
```
Comment (mirii1994) [inline grafana_objects/client_grafana_objects_delete_by_uid.go:16]: Whenever you initiate a call to the Logzio api, please use the CallLogzioApi function instead.
You can see an example here:
https://github.com/logzio/logzio_terraform_client/blob/master/archive_logs/client_archive_logs_retrieve.go#L19

File: grafana_objects/client_grafana_objects_delete_by_uid_test.go [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+	"io/ioutil"
+	"net/http"
+	"path"
+	"testing"
+
+	"github.com/logzio/logzio_terraform_client/grafana_objects"
+	"github.com/stretchr/testify/assert"
+)
+
+func deleteMockHandler(t *testing.T) func(http.ResponseWriter, *http.Request) {
```
Comment (mirii1994) [inline grafana_objects/client_grafana_objects_delete_by_uid_test.go:15]: Please refactor the test like in here:
https://github.com/logzio/logzio_terraform_client/blob/master/archive_logs/archive_logs_setup_test.go

Also, please add tests for responses 500, 404 from the server (you can use the archive logs dir as an example for how to do that).

File: grafana_objects/client_grafana_objects_get_by_uid.go [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+	"github.com/logzio/logzio_terraform_client/client"
+)
+
+const (
+	grafanaObjectsDashboardsByUID = grafanaObjectServiceEndpoint + "/uid/%s"
+)
+
+// Get allows getting a Grafana objects configuration.
+// https://docs.logz.io/api/#operation/getDashboarById
+func (c *GrafanaObjectsClient) Get(objectUid string) (*GetResults, error) {
```
Comment (mirii1994) [inline grafana_objects/client_grafana_objects_get_by_uid.go:19]: Whenever you initiate a call to the Logzio api, please use the CallLogzioApi function instead.
You can see an example here:
https://github.com/logzio/logzio_terraform_client/blob/master/archive_logs/client_archive_logs_retrieve.go#L19

File: grafana_objects/client_grafana_objects_get_by_uid_test.go [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+	"net/http"
+	"path"
+	"testing"
+
+	"github.com/logzio/logzio_terraform_client/grafana_objects"
+	"github.com/stretchr/testify/assert"
+)
+
+func getMockHandler(t *testing.T) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
```
Comment (mirii1994) [inline grafana_objects/client_grafana_objects_get_by_uid_test.go:16]: Please refactor the test like in here:
https://github.com/logzio/logzio_terraform_client/blob/master/archive_logs/archive_logs_setup_test.go

Also, please add tests for responses 500, 404 from the server (you can use the archive logs dir as an example for how to do that).

**Review by mirii1994 — CHANGES_REQUESTED** [review]
Hi @etiennejournet great progress, see my comments :)

File: grafana_objects/client_grafana_dashboard_structs.go (hunk 6 above)
Comment (etiennejournet) [reply grafana_objects/client_grafana_dashboard_structs.go:16]: Hello @mirii1994, I found those in https://grafana.com/docs/grafana/latest/dashboards/json-model/

File: grafana_objects/client_grafana_objects_delete_by_uid.go (hunk 7 above)
Comment (etiennejournet) [reply grafana_objects/client_grafana_objects_delete_by_uid.go]: done

File: grafana_objects/client_grafana_objects_delete_by_uid_test.go (hunk 8 above)
Comment (etiennejournet) [reply grafana_objects/client_grafana_objects_delete_by_uid_test.go]: done

File: grafana_objects/client_grafana_objects_get_by_uid.go (hunk 9 above)
Comment (etiennejournet) [reply grafana_objects/client_grafana_objects_get_by_uid.go:20]: done

File: grafana_objects/client_grafana_objects_get_by_uid_test.go (hunk 10 above)
Comment (etiennejournet) [reply grafana_objects/client_grafana_objects_get_by_uid_test.go]: done

**Comment by mirii1994** [comment]
@etiennejournet Great work, thank you!
I'll go over it & make some minor changes where needed, and add our integration tests :)

**Comment by mirii1994** [comment]
Great work @etiennejournet !
I'm merging it into a branch that will contain all the features and updates of version 1.11.
This will also allow running the tests (since it's a PR from a fork, the secrets that hold the env vars that required to run the tests are not available).

**Review by mirii1994 — APPROVED** [review]

**Comment by etiennejournet** [comment]
Hi @mirii1994, Working on the terraform side I figured some of the struct fields are not correct, some changes are needed because of unclear documentation, and personal mistake.

I have done those changes on my side, how do you want me to submit it to the new release ? Open a PR to 1.11.0 ?

**Comment by mirii1994** [comment]
@etiennejournet Yeah, open a PR to 1.11 :)
